### PR TITLE
Refactor components into a subclass.

### PR DESCRIPTION
**Pull request by dblock**
Signed-off-by: dblock <dblock@dblock.org>

### Description

Refactor a collection of components into a `Components` and `ComponentsManifest` base classes and use it everywhere. Makes it a lot easier to make a manifest with components that can be selected based on input options.
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/manifests/manifest.py [hunk 1]
```diff
@@ -51,15 +51,17 @@ def from_path(cls, path):
...
-                if nested:
-                    result[k] = nested
-            elif v and v != []:
-                result[k] = v
-        return result
+        if isinstance(d, list):
+            return list(map(lambda i: cls.compact(i), d))
+        elif isinstance(d, dict):
+            result = {}
+            for k, v in d.items():
```
Comment (owaiskazi19) [inline src/manifests/manifest.py:58]: Can we be more specific with `k` and `v` here and for the rest below?

File: src/manifests/manifest.py [hunk 2]
```diff
@@ -51,15 +51,17 @@ def from_path(cls, path):
...
+                v = cls.compact(v)
+                if v and v != []:
```
Comment (owaiskazi19) [inline src/manifests/manifest.py:60]:
```suggestion
                if v:
```
should do the work here

**Review by peternied — APPROVED** [review]

**Review by VachaShah — APPROVED** [review]

File: src/manifests/manifest.py (hunk 1 above)
Comment (dblock) [reply src/manifests/manifest.py:58]: Well, they are key and value, but nothing specific :)

File: src/manifests/manifest.py (hunk 2 above)
Comment (dblock) [reply src/manifests/manifest.py:60]: I wanted to capture intent. Is it true that an empty list returns false on []? Will check in a future PR. Weird.

Comment (dblock) [reply src/manifests/manifest.py:60]: https://github.com/opensearch-project/opensearch-build/pull/877

**Pull request merged**